Wrap gl texture

vlj opened the pull request:
This PR wraps OpenGL texture in an opaque structure so that it can be later used as an abstraction.
The next step will be to port buffer and then pipeline (alpha mode, shaders, ...)

perim commented on lib/ivis_opengl/gfx_api_gl.cpp:
```diff
@@ -0,0 +1,87 @@
+#include <algorithm>
+#include <cmath>
+
+namespace gl_api
```
Not a big fan of using namespaces in c++ files like this. Both because it clashes with existing style, causing unnecessary style changes when touching code, and because it leads to a lot of unnecessary indentation in general. Please use the <namespace>::<function name> declaration style instead (as you already do in the last function).

perim commented on lib/ivis_opengl/gfx_api_gl.cpp:
```diff
@@ -0,0 +1,87 @@
+
+namespace gl_api
+{
+	auto to_gl(const gfx_api::pixel_format& format)
```
auto return type is a c++14 feature, and we haven't enabled that yet in our build systems. Is there a strong reason to use it here?

perim commented on lib/ivis_opengl/gfx_api_gl.cpp:
```diff
@@ -0,0 +1,87 @@
+			for (unsigned i = 0; i < floor(log(std::max(width, height))); ++i)
+			{
+				glTexImage2D(GL_TEXTURE_2D, i, to_gl(internal_format), width >> i, height >> i, 0, to_gl(internal_format), GL_UNSIGNED_BYTE, nullptr);
+			}
```
I wonder if we should just take the step up to glTexStorage2D() here instead. It is only core in OpenGL 4.0, but should be in most drivers that don't support 4.0 too, I think.

perim commented on lib/ivis_opengl/gfx_api_gl.cpp:
```diff
@@ -0,0 +1,87 @@
+
+	struct context : public gfx_api::context
+	{
+		virtual std::unique_ptr<gfx_api::texture> create_texture(const size_t & width, const size_t & height, const gfx_api::pixel_format & internal_format, const std::string& filename) override
```
Why use unique_ptr here? (We use explicit resource creation and destruction, and clear ownership rather than magic pointers, in the rest of the code.)

perim commented on src/texture.cpp:
```diff
@@ -117,6 +118,9 @@ bool texLoad(const char *fileName)
 	firstPage = pie_NumberOfPages();
 
+	if (firstPage == 5)
+		printf("here");
```
Maybe remove this? :-)

vlj commented:
I updated the PR to implement your suggestions.
I'm still reluctant to use glTexStorage atm ; I'm trying to stick as much as possible with current function calls to avoid introducing regressions. However I'm intending to use it later when adding support for pre generated mipmap levels.

perim commented on lib/ivis_opengl/tex.cpp:
```diff
@@ -1,272 +1,298 @@
+	case 4:
+		return gfx_api::pixel_format::rgba;
+	default:
+		debug(LOG_ERROR, "iV_getPixelFormat: Unsupported image depth: %u", image->depth);
```
Maybe LOG_FATAL here, too, for consistency?

vlj commented:
I added debug(LOG_FATAL and fixed the dos line ending (there's bash for Windows however it doesn't tell me which file have wrong line ending).

What is the flag to pass to gcc to display missing case in switch statements ? For some reasons MXE doesn't show them by default (with cmake).

perim commented:
-Wswitch should give you warnings for missing switch enum statements. However, this should be included in -Wall already.

vlj commented:
I fixed some issues that appeared when removing std::unique_ptr. Should be good for another review.

perim commented:
Merged. Thanks!

past-due commented:
As a heads-up (in case the comment posted on the commit isn't seen):
This patch replaces `gluBuild2DMipmaps` with `glGenerateMipmap`, which is only available in OpenGL 3.0+. This is a breaking change, and may be part of why [this user suddenly has a crash with the latest buildbot build](http://forums.wz2100.net/viewtopic.php?f=1&p=137462#p137459).

vlj commented:
I'm investigating. I don't think glGenerateMipmap is the culprit here since the card is a 9500 which does support gl 3.0.
Maybe some texture isn't properly delete or used after delete.
